File: unsorted/java/273.integer-to-english-words.java
/*
 * @lc app=leetcode id=273 lang=java
 *
 * [273] Integer to English Words
 */

// @lc code=start
class Solution {
    public String numberToWords(int num) {
        // 2^31 - 1.
        // 4(billion),294(million)967(thousand)295
        if (num == 0)
            return "Zero";

        String[] arr = new String[4];
        for (int i = 0; i < 4; i++) {
            int curPart = num % 1000;
            arr[i] = translate(curPart);
            num = num / 1000;
        }
        String[] connections = { "", "Thousand", "Million", "Billion" };

        StringBuilder res = new StringBuilder();
        for (int i = 3; i >= 0; i--) {
            if (arr[i] != null && arr[i].length() != 0) {
                if (res.length() != 0)
                    res.append(" ");
                res.append(arr[i]);
                if (i != 0) {
                    res.append(" ");
                    res.append(connections[i]);
                }
            }
        }
        return res.toString();
    }

    String translate(int num) {
        // num <1000
        String[] bitsArray = { "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine" };
        String[] tensArray = { "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety" };
        String[] basicBitsArray = { "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
                "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen" };
        StringBuilder res = new StringBuilder();
        int hundredPlace = num / 100;
        if (hundredPlace != 0) {
            res.append(bitsArray[hundredPlace]);
            res.append(" Hundred");
            num = num - hundredPlace * 100;
        }
        int tensPlace = num / 10;
        if (tensPlace >= 2) {
            if (res.length() != 0)
                res.append(" ");
            res.append(tensArray[tensPlace]);

            num = num - tensPlace * 10;
        }
        if (num != 0) {
            if (res.length() != 0)
                res.append(" ");
            res.append(basicBitsArray[num]);
        }
        return res.toString();
    }
}
// @lc code=end

File: js/leetcode/js/273.integer-to-english-words.js
/*
 * @lc app=leetcode id=273 lang=javascript
 *
 * [273] Integer to English Words
 */

// @lc code=start
/**
 * @param {number} num
 * @return {string}
 */
var numberToWords = function (num) {
    let setMap = ['', 'Thousand', 'Million', 'Billion']
    let curSetPos = 0;
    let res = ''
    if (num === 0)
        return 'Zero'
    while (num) {
        const curSet = num % 1000
        const curRes = integerToText(curSet)
        if (curRes)
            res = curRes + (setMap[curSetPos] && ' ') + setMap[curSetPos] + (setMap[curSetPos] && ' ') + res
        num = Math.floor(num / 1000)
        curSetPos++;
    }
    return res;
    //num < 1000
    function integerToText(num) {
        const hundred = ['', 'One', 'Two', 'Three', 'Four',
            'Five', 'Six', 'Seven', 'Eight', 'Nine'
        ]
        const tenthGreaterThanTen = ['', '', 'Twenty', 'Thirty', 'Forty', 'Fifty',
            'Sixty', 'Seventy', 'Eighty', 'Ninety']
        const tenth = ['Ten', 'Eleven', 'Twelve', 'Thirteen', 'Fourteen',
            'Fifteen', 'Sixteen', 'Seventeen', 'Eighteen', 'Nineteen']
        const last = hundred;

        let res = ''
        const hundredsBit = Math.floor(num / 100);
        const tensBit = Math.floor(Math.floor(num % 100) / 10)
        const lastBit = Math.floor(num % 10)

        if (hundredsBit)
            res = res + hundred[hundredsBit] + ' Hundred '
        if (tensBit === 1) {
            res += tenth[lastBit]
        } else {
            const tenRes = tenthGreaterThanTen[tensBit]
            if (tenRes)
                res = res + tenthGreaterThanTen[tensBit]
            res = res + (last[lastBit] && ' ') + last[lastBit]
        }
        return res;
    }
}

// console.log(numberToWords(12345))

// @lc code=end

